src/Traits/GenericActionsTrait.php

```diff
@@ -57,23 +57,119 @@ protected function loadFilteredChildren(string $id): array
+        if ($event->getResult() !== null) {
+            return $event->getResult();
+        }
+
+        $event = $this->dispatchEvent('Controller.beforeObjectLoad', compact('uname', 'type'));
+        if ($event->getResult() !== null) {
+            return $event->getResult();
+        }
+
+        return null;
+    }
+
+    /**
+     * Dispatch after object load hook events.
+     *
+     * @param string $uname Object uname.
+     * @param string $type Object type.
+     * @param array $data Objects data to pass to callback.
+     * @return \Cake\Http\Response|null The resulting response of the events.
+     */
+    protected function dispatchAfterLoadEvent(string $uname, string $type, array $data): ?Response
+    {
+        $event = $this->dispatchEvent(sprintf('Controller.afterObjectLoad:%s', $uname), $data);
+        if ($event->getResult() !== null) {
+            return $event->getResult();
+        }
+
+        $event = $this->dispatchEvent(sprintf('Controller.afterObjectLoad:%s', $type), $data);
+        if ($event->getResult() !== null) {
+            return $event->getResult();
+        }
+
+        $event = $this->dispatchEvent('Controller.afterObjectLoad', $data);
+        if ($event->getResult() !== null) {
+            return $event->getResult();
+        }
+
+        return null;
+    }
+
+    /**
+     * Dispatch before object render hook events.
+     *
+     * @param string $uname Object uname.
+     * @param string $type Object type.
+     * @param array $data Objects data to pass to callback.
+     * @return \Cake\Http\Response|null The resulting response of the events.
+     */
+    protected function dispatchBeforeRenderEvent(string $uname, string $type, array $data): ?Response
+    {
+        $event = $this->dispatchEvent(sprintf('Controller.beforeObjectRender:%s', $uname), $data);
+        if ($event->getResult() !== null) {
+            return $event->getResult();
+        }
+
+        $event = $this->dispatchEvent(sprintf('Controller.beforeObjectRender:%s', $type), $data);
+        if ($event->getResult() !== null) {
+            return $event->getResult();
+        }
+
+        return null;
+    }
+
     /**
      * Generic objects route.
      *
-     * @param string $id Object id
+     * @param string $uname Object id or uname.
      * @return Response
      */
-    public function objects(string $id): Response
+    public function objects(string $uname): Response
     {
-        $object = $this->Objects->loadFullObject($id);
-        if ($object->type === 'folders') {
-            $children = $this->loadFilteredChildren($object->uname);
-            $object['children'] = $children;
-            $this->set(compact('children'));
+        $object = $this->Objects->loadObject($uname, 'objects', [], []);
+        $result = $this->dispatchBeforeLoadEvent($object->uname, $object->type);
+        if ($result !== null) {
+            return $result;
```

Comment:
Doing so would mean that the listeners are responsible for returning a valid `Response` object.

Are we sure it is what we need, especially for the `beforeLoad` and `afterLoad` events?
I get that this makes sense for the `beforeObjectRender` event, but I'd argue it is not what I would expect from the other two:

- `Controller.beforeObjectLoad`: either an array of options to be passed to the loader, or an object entity — in which case, the load is skipped
- `Controller.afterObjectLoad`:  a modified object entity
- `Controller.beforeObjectRender`: a Response

Also, in the methods responsible for dispatching the events, I'd assert that the result, when not null, is of the expected type, and if not I'd throw an `UnexpectedValueException` with a helpful message, otherwise it may be hard to debug the problem if the listener doesn't return a valid result for the event.

Reply:
I thought that returning a `Response` was the most versatile solution. This way, whatever the hook is, we can return a custom template or load objects differently depending on the object itself and whose route we do not know regardless. For example, we would paging the `composed_by` relation of all requested `documents` using the `document_collections` layout if the relation has items. Or even make special redirects in case the requested detail is of type `link`.